packages/vitest/src/node/optimizedGrouper.ts

```diff
@@ -0,0 +1,696 @@
+/**
+ * Optimized Test Grouping Algorithm
+ *
+ * This module provides advanced test grouping strategies that optimize for:
+ * 1. Minimal total execution time (load balancing)
+ * 2. Early failure detection (failed tests first)
+ * 3. Resource utilization (bin packing)
+ * 4. Dependency awareness (minimize shared resource contention)
+ * 5. Cache efficiency (co-locate related tests)
+ */
+
+import type { SuiteResultCache } from './cache/results'
+import type { TestSpecification } from './spec'
+
+export interface TestMetadata {
+  spec: TestSpecification
+  duration: number // milliseconds
+  fileSize: number // bytes
+  failed: boolean
+  dependencies: Set<string> // module dependencies
+  testCount?: number
+  environment?: string // test environment (node, jsdom, etc)
+}
+
+export interface OptimizedGroup {
+  specs: TestSpecification[]
+  estimatedDuration: number
+  priority: number // lower = runs earlier
+  poolType: string
+  projectName: string
+  environment?: string // test environment
+}
+
+export interface GroupingStrategy {
+  name: string
+  group: (tests: TestMetadata[], maxWorkers: number) => OptimizedGroup[][]
+}
+
+/**
+ * Main optimized test grouper
+ */
+export class OptimizedTestGrouper {
+  private resultsCache: Map<string, SuiteResultCache>
+  private filesCache: Map<string, { size: number }>
+
+  constructor(
+    resultsCache: Map<string, SuiteResultCache>,
+    filesCache: Map<string, { size: number }>,
+  ) {
+    this.resultsCache = resultsCache
+    this.filesCache = filesCache
+  }
+
+  /**
+   * Enriches test specifications with metadata for optimization
+   */
+  enrichMetadata(specs: TestSpecification[], environments?: WeakMap<TestSpecification, { name: string; options: any }>): TestMetadata[] {
+    return specs.map((spec) => {
+      // Build cache key: "projectName:relativePath"
+      const cacheKey = `${spec.project.name}:${spec.moduleId}`
+      const cached = this.resultsCache.get(cacheKey)
+      const fileStats = this.filesCache.get(cacheKey)
+      const fileSize = fileStats?.size || 0
+
+      return {
+        spec,
+        duration: cached?.duration || this.estimateDuration(fileSize),
+        fileSize,
+        failed: cached?.failed || false,
+        dependencies: this.extractDependencies(spec),
+        testCount: undefined, // Can be populated if available
+        environment: environments?.get(spec)?.name,
+      }
+    })
+  }
+
+  /**
+   * Estimates duration based on file size (heuristic: 1ms per 100 bytes)
+   */
+  private estimateDuration(fileSize: number): number {
+    // Base assumption: larger files take longer
+    // Minimum 10ms, scales with file size
+    return Math.max(10, fileSize / 100)
+  }
+
+  /**
+   * Extracts module dependencies from test specification
+   */
+  private extractDependencies(_spec: TestSpecification): Set<string> {
+    // This would ideally use Vite's module graph
```

Comment:
It wouldn't be populated at this point though? Only in watch mode re-runs. And even then we don't automatically re-run the tests where module graph didn't change.